File: deno/src/base/tenant/tenant.service.ts
import {
  findAllTenant,
  delCacheTenant,
} from "/gen/base/tenant/tenant.dao.ts";

import {
  findAllDomain,
  delCacheDomain,
} from "/gen/base/domain/domain.dao.ts";

import type {
  GetLoginTenants,
  SetTenantAdminPwdInput,
} from "/gen/types.ts";

import {
  findByIdsTenant,
  findByIdTenant,
  validateOptionTenant,
} from "/gen/base/tenant/tenant.dao.ts";

import {
  findByIdLang,
} from "/gen/base/lang/lang.dao.ts";

import {
  findOneUsr,
  createUsr,
  updateByIdUsr,
} from "/gen/base/usr/usr.dao.ts";

export async function getLoginTenants(
  domain: string,
): Promise<GetLoginTenants[]> {
  let domain_models = await findAllDomain({
    lbl: domain,
    is_enabled: [ 1 ],
  });
  if (domain_models.length === 0) {
    await delCacheDomain();
    domain_models = await findAllDomain({
      lbl: domain,
      is_enabled: [ 1 ],
    });
  }
  const res: GetLoginTenants[] = [ ];
  if (domain_models.length > 0) {
    let domain_ids: DomainId[] = domain_models.map((item) => item.id);
    let tenant_models = await findAllTenant({
      domain_ids,
      is_enabled: [ 1 ],
    });
    if (tenant_models.length === 0) {
      await delCacheDomain();
      domain_models = await findAllDomain({
        lbl: domain,
        is_enabled: [ 1 ],
      });
      domain_ids = domain_models.map((item) => item.id);
      await delCacheTenant();
      tenant_models = await findAllTenant({
        domain_ids,
        is_enabled: [ 1 ],
      });
    }
    for (const tenant_model of tenant_models) {
      const lang_id = tenant_model.lang_id;
      const lang_model = await findByIdLang(lang_id);
      const lang = lang_model?.code || "zh-CN";
      res.push({
        id: tenant_model.id,
        lbl: tenant_model.lbl,
        title: tenant_model.title,
        info: tenant_model.info,
        lang,
      });
    }
  }
  return res;
}

/** 根据 租户ids 获取 租户信息 */
export async function getLoginTenantByIds(
  tenant_ids: TenantId[],
): Promise<GetLoginTenants[]> {
  const tenant_models = await findByIdsTenant(tenant_ids);
  const res: GetLoginTenants[] = [ ];
  for (const tenant_model of tenant_models) {
    const lang_id = tenant_model.lang_id;
    const lang_model = await findByIdLang(lang_id);
    const lang = lang_model?.code || "zh-CN";
    res.push({
      id: tenant_model.id,
      lbl: tenant_model.lbl,
      title: tenant_model.title,
      info: tenant_model.info,
      lang,
    });
  }
  return res;
}

/** 设置租户管理员密码 */
export async function setTenantAdminPwd(
  input: SetTenantAdminPwdInput,
): Promise<boolean> {
  const tenant_id = input.tenant_id;
  const pwd = input.pwd;
  await validateOptionTenant(
    await findByIdTenant(tenant_id),
  );
  const usr_model = await findOneUsr({
    username: "admin",
    tenant_id,
  });
  if (!usr_model) {
    await createUsr({
      username: "admin",
      password: pwd,
      tenant_id,
    });
  } else {
    await updateByIdUsr(
      usr_model.id,
      {
        password: pwd,
      },
    );
  }
  return true;
}

File: rust/src/src/base/tenant/tenant_service.rs
use color_eyre::eyre::Result;
// use crate::common::context::get_auth_tenant_id;

use crate::r#gen::base::tenant::tenant_dao::{
  find_all as find_all_tenant,
  del_cache as del_cache_tenant,
};
use crate::r#gen::base::tenant::tenant_model::TenantSearch;

use super::tenant_model::GetLoginTenants;

use crate::r#gen::base::domain::domain_dao::{
  find_all as find_all_domain,
  del_cache as del_cache_domain,
};
use crate::r#gen::base::domain::domain_model::{
  DomainSearch,
  DomainModel,
  DomainId,
};

// 租户
use crate::r#gen::base::tenant::tenant_dao::{
  find_by_id as find_by_id_tenant,
  validate_option as validate_option_tenant,
};

// 语言
use crate::r#gen::base::lang::lang_dao::find_by_id as find_by_id_lang;

// 用户
use crate::r#gen::base::usr::usr_dao::{
  find_one as find_one_usr,
  update_by_id as update_by_id_usr,
  create as create_usr,
};
use crate::r#gen::base::usr::usr_model::{
  UsrInput,
  UsrSearch,
};

use super::tenant_model::SetTenantAdminPwdInput;

// 获取当前租户绑定的网址
// pub async fn get_host_tenant() -> Result<String> {
  
//   let tenant_id = get_auth_tenant_id();
  
//   if tenant_id.is_none() {
//     return Err(eyre!("tenant_id is none"));
//   }
  
//   let tenant_id = tenant_id.unwrap();
  
//   let model = tenant_dao::find_one(
//     TenantSearch {
//       id: tenant_id.into(),
//       ..Default::default()
//     }.into(),
//     None,
//     None,
//   ).await?;
  
//   if model.is_none() {
//     return Err(eyre!("tenant_id cannot be found"));
//   }
  
//   let model = model.unwrap();
  
//   let domain = model.domain;
  
//   Ok(domain)
// }

/// 根据 当前网址的域名+端口 获取 租户列表
pub async fn get_login_tenants(
  domain: String,
) -> Result<Vec<GetLoginTenants>> {
  
  let mut domain_models: Vec<DomainModel> = find_all_domain(
    DomainSearch {
      lbl: domain.clone().into(),
      is_enabled: vec![1].into(),
      ..Default::default()
    }.into(),
    None,
    None,
    None,
  ).await?;
  
  if domain_models.is_empty() {
    del_cache_domain().await?;
    domain_models = find_all_domain(
      DomainSearch {
        lbl: domain.into(),
        is_enabled: vec![1].into(),
        ..Default::default()
      }.into(),
      None,
      None,
      None,
    ).await?;
  }
  
  let domain_ids: Vec<DomainId> = domain_models
    .into_iter()
    .map(|x| x.id)
    .collect();
  
  if domain_ids.is_empty() {
    return Ok(vec![]);
  }
  
  let mut tenant_models = find_all_tenant(
    TenantSearch {
      domain_ids: domain_ids.clone().into(),
      is_enabled: vec![1].into(),
      ..Default::default()
    }.into(),
    None,
    None,
    None,
  ).await?;
  
  if tenant_models.is_empty() {
    del_cache_domain().await?;
    del_cache_tenant().await?;
    tenant_models = find_all_tenant(
      TenantSearch {
        domain_ids: domain_ids.into(),
        is_enabled: vec![1].into(),
        ..Default::default()
      }.into(),
      None,
      None,
      None,
    ).await?;
  }
  
  let mut res: Vec<GetLoginTenants> = vec![];
  
  for tenant_model in tenant_models {
    
    let lang_model = find_by_id_lang(
      tenant_model.lang_id.clone(),
      None,
    ).await?;
    
    let lang = lang_model
      .map(|x| x.code)
      .unwrap_or_else(|| "zh-CN".to_owned());
    
    res.push(GetLoginTenants {
      id: tenant_model.id,
      lbl: tenant_model.lbl,
      lang,
    });
    
  }
  
  Ok(res)
}

/// 设置租户管理员密码
pub async fn set_tenant_admin_pwd(
  input: SetTenantAdminPwdInput,
) -> Result<bool> {
  
  let tenant_id = input.tenant_id;
  let pwd = input.pwd;
  
  let tenant_model = find_by_id_tenant(
    tenant_id.clone(),
    None,
  ).await?;
  validate_option_tenant(tenant_model).await?;
  
  let usr_model = find_one_usr(
    UsrSearch {
      username: "admin".to_owned().into(),
      tenant_id: tenant_id.clone().into(),
      ..Default::default()
    }.into(),
    None,
    None,
  ).await?;
  
  if let Some(usr_model) = usr_model {
    update_by_id_usr(
      usr_model.id,
      UsrInput {
        password: Some(pwd),
        ..Default::default()
      },
      None,
    ).await?;
  } else {
    create_usr(
      UsrInput {
        username: "admin".to_owned().into(),
        password: pwd.to_owned().into(),
        tenant_id: tenant_id.into(),
        ..Default::default()
      },
      None,
    ).await?;
  }
  
  Ok(true)
}
